File: tools/dsl-harness/kernels/triton/py/vector_exp_bwd.py
"""Triton vector-exp backward kernel: grad_input = grad_output * output."""

from __future__ import annotations

from typing import Any, Dict, Tuple

import triton
import triton.language as tl

from triton_helpers import fake


@triton.jit
def triton_exp_backward_kernel(
    grad_output_ptr,
    output_ptr,
    grad_input_ptr,
    n_elements,
    BLOCK_SIZE: tl.constexpr,
):
    pid = tl.program_id(axis=0)
    block_start = pid * BLOCK_SIZE
    offsets = block_start + tl.arange(0, BLOCK_SIZE)
    mask = offsets < n_elements
    grad_output = tl.load(grad_output_ptr + offsets, mask=mask)
    output = tl.load(output_ptr + offsets, mask=mask)
    grad_input = grad_output * output
    tl.store(grad_input_ptr + offsets, grad_input, mask=mask)


def build_args(cfg: Dict[str, Any]) -> Tuple[list, Dict[str, Any]]:
    return [
        fake("fp32", 1024),  # grad_output_ptr
        fake("fp32", 1024),  # output_ptr
        fake("fp32", 1024),  # grad_input_ptr
        1024,                # n_elements
    ], {
        "BLOCK_SIZE": int(cfg.get("BLOCK_SIZE", 1024)),
        "num_warps": 4,
        "num_stages": 1,
    }

File: tools/dsl-harness/kernels/triton/zig/vector_exp_bwd.zig
//! Registration of `triton_exp_backward_kernel` (gradient of vector exp).

const std = @import("std");

const harness = @import("harness");
const zml = @import("zml");
const ops = zml.ops;
const tri = zml.kernel.triton;
const Tensor = zml.Tensor;

const Cfg = struct { BLOCK_SIZE: i32 = 1024 };

pub const Kernel = tri.Kernel(Cfg, .{
    .name = "triton_exp_backward_kernel",
    .inputs = &.{ "grad_output_ptr", "output_ptr", "n_elements" },
    .outputs = &.{"grad_input"},
    .run = run,
});

fn run(b: *tri.Builder, cfg: Cfg) tri.FinishError!void {
    const a = try b.declareArgs(.{
        .grad_output_ptr = .{ .ptr = .f32 },
        .output_ptr = .{ .ptr = .f32 },
        .grad_input_ptr = .{ .ptr = .f32 },
        .n_elements = .{ .scalar_opts = .{ .dtype = .i32, .divisibility = 16 } },
    });

    const pid = b.programId(.x);
    const offsets = pid.mul(cfg.BLOCK_SIZE).add(b.arange(0, cfg.BLOCK_SIZE, .i32));
    const mask = offsets.lt(a.n_elements);

    const grad_output = b.loadOpts(a.grad_output_ptr.addPtr(offsets), .{ .mask = mask });
    const output = b.loadOpts(a.output_ptr.addPtr(offsets), .{ .mask = mask });
    const grad_input = grad_output.mul(output);
    b.storeOpts(a.grad_input_ptr.addPtr(offsets), grad_input, .{ .mask = mask });
}

pub const SWEEPS: []const harness.Sweep(Kernel.Config) = &.{
    .{ .label = "default", .cfg = .{ .BLOCK_SIZE = 1024 } },
};

threadlocal var active_ttir: [:0]const u8 = "";

pub fn setActiveTtir(ttir: [:0]const u8) void {
    active_ttir = ttir;
}

pub fn forward(grad_out: Tensor, out: Tensor, _: Tensor, n: Tensor) Tensor {
    return ops.triton(
        .{ grad_out, out, n },
        .{grad_out.shape()},
        .{
            .name = Kernel.name,
            .ir = active_ttir,
            .grid = .{ 1, 1, 1 },
            .num_warps = 4,
            .num_stages = 1,
        },
    )[0];
}

pub fn args() std.meta.ArgsTuple(@TypeOf(forward)) {
    return .{
        Tensor.init(.{1024}, .f32), // grad_output_ptr
        Tensor.init(.{1024}, .f32), // output_ptr
        Tensor.init(.{1024}, .f32), // grad_input_ptr placeholder
        Tensor.init(.{}, .i32),     // n_elements
    };
}
